### PR TITLE
Adding `n_grid_points` parameter to rise/set/transit functions

### PR DESCRIPTION
**Pull request by bmorris3**
**tl;dr**: you can speed up rise/set times by an order of magnitude in exchange for a few minutes of rise/set time precision, without changing anything fundamental about how astroplan/astropy compute rise/set times.

This PR implements the changes I suggested in #423, which I summarize briefly below. There are also several PEP8 compliance fixes throughout the docstrings on the Observer object. If merged, this PR supersedes #327.

The primary goal of this PR is to expose the `N` parameter, which has been lurking within the guts of the Observer object since its inception. Rise/set/transit times are computed using a grid search, which scans for horizon crossings over `N=150` times spread over 24 hours relative to the reference time supplied by the user. By throttling that parameter up or down, the user can simultaneously change the precision and speed of rise/set/transit calculations. `N~100` gives precision better than 1 min from the true rise/set/transit time, `N~10` gives precision better than 5 minutes and is about 10x faster. This fix is designed to address the common complaint that rise/set times are slow (#402, #423).

The default behavior of the functions remains unchanged. When not supplied by the user, `N=150` by default for all functions. Thus none of the tests need to change, and introducing this new optional kwarg won't change people's results in legacy code.

Still to do: 
* ~add narrative docs explaining the update~
* include plots in docs comparing precision at different `N`

### CODE REVIEW

**Comment by bmorris3** [comment]
Tagging some users who might have opinions about this implementation: @wtgee, @StuartLittlefair, @iwellaway, @eteq.

**Comment by StuartLittlefair** [comment]
@bmorris3 I like this approach a lot, and it does indeed make things a lot faster, and is simple to understand for the end user. 

For a single time/target I find the speedup between ```N=150``` and ```N=10``` is only x2, but this obviously improves with grids of times and targets, since ```N``` broadcasts against those grids, so the speedup will be more impressive if doing large numbers of calculations.

One thing I've often wondered is why we couldn't use a 2-pass scenario, where we use a small ```N``` to get an initial rough estimate of the event time, and then re-grid 10--15 mins either side of this estimate (using the same ```N```) to refine it, i.e use a few lines like the ones below at the end of these functions...

```python
    t = self._two_point_interp(jd1, jd2, al1, al2, horizon=horizon)
    # normally we'd return here, but let's use a finer grid, 
    # 15 mins either side of rough time
    times = _generate_24hr_grid(t, -0.01, 0.01, N)
    altaz = self.altaz(times, target, grid_times_targets=grid_times_targets)
    altitudes = altaz.alt
    al1, al2, jd1, jd2 = self._horiz_cross(times, altitudes, rise_set, horizon)
    return self._two_point_interp(jd1, jd2, al1, al2, horizon=horizon)
```

This is roughly the same speed for a single time/target for ```N=10``` vs ```N=150``` (as the two calculations offset the x2 speedup), but is still 5x faster for a grid of 300 times and agrees with the ```N=150``` case to within 1 second.

Or - giving the option to the user, you could switch a two-pass behaviour on or off using an argument, and allow the user to tune ```N``` and ```two_pass``` for their own accuracy requirements...

**Comment by bmorris3** [comment]
@StuartLittlefair I seem to remember that it was originally our intention to use a two-pass solution, but looking back through the PR history, I can't find evidence that we ever actually did a two-pass solution. 

I think something @eteq and I discussed casually a few times is that atmospheric pressure and humidity along the line of sight can probably change rise/set times by a few minutes here or there, so we wondered whether it would be physically meaningful to provide rise/set times that are accurate to the minute anyways. I suppose I'm making the arguments that (1) perhaps the default `N` could be smaller, and (2) a two-pass solution is probably overkill.

**Comment by StuartLittlefair** [comment]
I don't really disagree with any of that. But I would make the following counter-arguments:

1) for *any* given accuracy requirement, a two-pass solution is quicker than just using a finer grid over 24 hour

2) It seems like a simple change to the code, I think the lines pasted above work fine

**Comment by bmorris3** [comment]
I've updated the PR to include narrative docs, which render like so: 

Screenshot 2019-10-24 at 11 32 56

**Comment by StuartLittlefair** [comment]
Nice. I don't think my comments should hold up merging this PR BTW - I could always submit a separate PR for a two-pass approach at a later date.

**Comment by tepickering** [comment]
this looks good and definitely helps in cases where precision isn't too important (like twilight times). i think there's still more to be gained from either a two-pass approach as suggested by @StuartLittlefair or an iterative approach like `skyfield` uses (see `find_discrete` in https://github.com/skyfielders/python-skyfield/blob/master/skyfield/almanac.py). those can be subjects for future PRs, though, and shouldn't hold this up...
